akka-stream/src/main/scala/akka/stream/scaladsl/Source.scala

```diff
@@ -561,6 +561,43 @@ object Source {
       case akka.actor.Status.Success                        => CompletionStrategy.Draining
     }, { case akka.actor.Status.Failure(cause)              => cause }, bufferSize, overflowStrategy)
 
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] def actorRefWithAck[T](
+      ackMessage: Any,
+      completionMatcher: PartialFunction[Any, CompletionStrategy],
+      failureMatcher: PartialFunction[Any, Throwable]): Source[T, ActorRef] = {
+    Source
+      .fromGraph(new ActorRefBackpressureSource(ackMessage, completionMatcher, failureMatcher))
+      .withAttributes(DefaultAttributes.actorRefWithAckSource)
+  }
+
+  /**
+   * Creates a `Source` that is materialized as an [[akka.actor.ActorRef]].
+   * Messages sent to this actor will be emitted to the stream if there is demand from downstream,
+   * and a new message will only be accepted after the previous messages has been consumed and acknowledged back.
+   *
+   * The stream can be completed successfully by sending the actor reference a [[akka.actor.Status.Success]].
+   * If the content is [[akka.stream.CompletionStrategy.immediately]] the completion will be signaled immidiately,
+   * otherwise if the content is [[akka.stream.CompletionStrategy.draining]] (or anything else)
+   * already buffered element will be signaled before siganling completion.
+   *
+   * The stream can be completed with failure by sending a [[akka.actor.Status.Failure]] to the
+   * actor reference. In case the Actor is still draining its internal buffer (after having received
+   * a [[akka.actor.Status.Success]]) before signaling completion and it receives a [[akka.actor.Status.Failure]],
+   * the failure will be signaled downstream immediately (instead of the completion signal).
+   *
+   * The actor will be stopped when the stream is completed, failed or canceled from downstream,
+   * i.e. you can watch it to get notified when that happens.
+   */
+  def actorRefWithAck[T](ackMessage: Any): Source[T, ActorRef] =
+    actorRefWithAck(ackMessage, {
+      case akka.actor.Status.Success(s: CompletionStrategy) => s
+      case akka.actor.Status.Success(_)                     => CompletionStrategy.Draining
+      case akka.actor.Status.Success                        => CompletionStrategy.Draining
+    }, { case akka.actor.Status.Failure(cause)              => cause })
```

Comment:
Given that the graph stage is written specifically for this use case, can we push these partial functions into the stage/actor instead, making it less generic/easer to reason about?

Reply:
I've implemented it similar to the plain`actorRef` API. `akka.stream.typed.scaladsl.ActorSource` is providing it with completion matcher. In case we also want to provide a similar API for `actorRefWithAck` I'd leave it at that.

Reply:
Ok, sounds like that makes sense, thanks for the clarification.